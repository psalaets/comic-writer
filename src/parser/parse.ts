import * as perf from '../perf';
import * as parts from '../comic-part-types';
import countWords from './count-words';
import * as classifiers from './line-classifiers';
import {
  Spread,
  Panel,
  Paragraph,
  Metadata,
  Dialogue,
  Caption,
  Sfx,
  LetteringContentChunk,
  BlankLine,
  PreSpread,
  SpreadChild,
  PanelChild,
  SpreadContent,
} from './types';

import {
  SPREAD_REGEX,
  PANEL_REGEX,
  CAPTION_REGEX,
  SFX_REGEX,
  DIALOGUE_REGEX,
  METADATA_REGEX,
  LETTERING_BOLD_REGEX
} from './regexes';

// flyweight pattern: reuse this to represent every blank line
const BLANK_LINE: BlankLine = {
  type: parts.BLANK
};

export const parseSpreadContent = perf.wrap('parseSpreadContent', parse);

function parse(content: SpreadContent): Spread<SpreadChild> {
  const spreadLine = content.spread;
  const childLines = content.children;

  const children = parseSpreadChildren(childLines);

  return parseSpread(spreadLine, children);
}

function parseSpreadChildren(lines: Array<string>): Array<SpreadChild> {
  const numbering = createLetteringNumberer();

  const spreadChildren: Array<SpreadChild> = [];
  let currentPanel: Panel<PanelChild> | null = null;

  for (let i = 0; i < lines.length; i++) {
    const parsed = parseSpreadChild(lines[i], numbering);

    if (parsed.type === parts.PANEL) {
      currentPanel = parsed;
      spreadChildren.push(parsed);
    } else {
      if (currentPanel) {
        currentPanel.children.push(parsed);
      } else {
        spreadChildren.push(parsed);
      }
    }
  }

  return applyPanelRollups(spreadChildren);
}

function* createLetteringNumberer() {
  let number = 1;
  while (true) {
    yield number++;
  }
}

function applyPanelRollups(children: Array<SpreadChild>): Array<SpreadChild> {
  children.forEach(spreadChild => {
    if (spreadChild.type === parts.PANEL) {
      const panel = spreadChild;

      panel.children.forEach(panelChild => {
        switch (panelChild.type) {
          case parts.CAPTION: {
            panel.captionCount += 1;
            panel.captionWordCount += panelChild.wordCount;
            break;
          }
          case parts.DIALOGUE: {
            panel.dialogueCount += 1;
            panel.dialogueWordCount += panelChild.wordCount;
            panel.speakers.push(panelChild.speaker);
            break;
          }
          case parts.SFX: {
            panel.sfxCount += 1;
            break;
          }
        }
      });
    }
  });

  return children;
}

function parseSpreadChild(line: string, numbering: Generator<number>): SpreadChild {
  // scripts are about 1/2 blank lines so this should be first
  if (classifiers.isBlank(line)) return BLANK_LINE;

  if (classifiers.isCaption(line)) return parseCaption(line, numbering);
  if (classifiers.isSfx(line)) return parseSfx(line, numbering);
  // dialogue has to be checked after sfx/caption, otherwise we get balloons
  // where the speaker is "caption" and "sfx"
  if (classifiers.isDialogue(line)) return parseDialogue(line, numbering);

  if (classifiers.isPanel(line)) return parsePanel(line);

  // any non-blank line can be a paragraph so it goes last
  return parseParagraph(line);
}

export function parsePreSpreadLines(lines: Array<string>): Array<PreSpread> {
  return lines.map(line => parsePreSpreadLine(line));
}

function parsePreSpreadLine(line: string): PreSpread {
  // scripts are about 1/2 blank lines so this should be first
  if (classifiers.isBlank(line)) return BLANK_LINE;

  if (classifiers.isMetadata(line)) return parseMetadata(line);

  // any non-blank line can be a paragraph so it goes last
  return parseParagraph(line);
}

export function parseSpread(line: string, children: Array<SpreadChild>): Spread<SpreadChild> {
  const matchResult = SPREAD_REGEX.exec(line) as Array<string>;

  const startPage = Number(matchResult[1]);
  const endPage = matchResult[3] != null ? Number(matchResult[3]) : startPage;
  const pageCount = countPages(startPage, endPage);

  return applySpreadRollups({
    type: parts.SPREAD,
    pageCount,
    children,

    // these start with default values, they're populated when applying rollups
    panelCount: 0,
    speakers: [],
    dialogueCount: 0,
    captionCount: 0,
    sfxCount: 0,
    dialogueWordCount: 0,
    captionWordCount: 0
  });
}

function applySpreadRollups(spread: Spread<SpreadChild>): Spread<SpreadChild> {
  return spread.children.reduce((spread, child) => {
    switch (child.type) {
      case parts.PANEL:
        spread.panelCount        += 1;
        spread.captionCount      += child.captionCount;
        spread.captionWordCount  += child.captionWordCount;
        spread.dialogueCount     += child.dialogueCount;
        spread.dialogueWordCount += child.dialogueWordCount;
        spread.sfxCount          += child.sfxCount;
        spread.speakers.push(...child.speakers);
        break;
      case parts.DIALOGUE:
        spread.dialogueCount     += 1;
        spread.dialogueWordCount += child.wordCount;
        break;
      case parts.CAPTION:
        spread.captionCount      += 1;
        spread.captionWordCount  += child.wordCount;
        break;
    }

    return spread;
  }, spread);
}

function countPages(startPage: number, endPage?: number): number {
  if (endPage == null) {
    return 1;
  } else if (startPage < endPage) {
    return (endPage - startPage) + 1;
  } else if (startPage > endPage) {
    return 2;
  } else { // startPage === endPage
    return 1;
  }
}

export function parsePanel(line: string): Panel<PanelChild> {
  const [, number] = PANEL_REGEX.exec(line) as Array<string>;

  return {
    type: parts.PANEL,
    number: Number(number),

    // these start with default values, they'll be populated later
    children: [],
    speakers: [],
    dialogueCount: 0,
    captionCount: 0,
    sfxCount: 0,
    dialogueWordCount: 0,
    captionWordCount: 0,
  };
}

export function parseMetadata(line: string): Metadata {
  const [, name, value] = METADATA_REGEX.exec(line) as Array<string>;

  return {
    type: parts.METADATA,
    name,
    value,
  };
}

export function parseParagraph(line: string): Paragraph {
  return {
    type: parts.PARAGRAPH,
    content: line
  };
}

export function parseDialogue(line: string, numbering: Generator<number>): Dialogue {
  const [, speaker, modifier, content] = DIALOGUE_REGEX.exec(line) as Array<string>;

  const parseTree = parseLetteringContent(content);

  return {
    type: parts.DIALOGUE,
    number: numbering.next().value,
    speaker,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content: parseTree,
    wordCount: countWords(parseTree),
  };
}

export function parseCaption(line: string, numbering: Generator<number>): Caption {
  const [, modifier, content] = CAPTION_REGEX.exec(line) as Array<string>;
  const parseTree = parseLetteringContent(content);

  return {
    type: parts.CAPTION,
    number: numbering.next().value,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content: parseTree,
    wordCount: countWords(parseTree)
  };
}

export function parseSfx(line: string, numbering: Generator<number>): Sfx {
  const [, modifier, content] = SFX_REGEX.exec(line) as Array<string>;

  return {
    type: parts.SFX,
    number: numbering.next().value,
    modifier: modifier ? modifier.slice(1, -1) : null,
    content,
  };
}

function parseLetteringContent(content: string): Array<LetteringContentChunk> {
  const chunks: Array<LetteringContentChunk> = [];

  let index = 0;
  let result = null;

  // eslint-disable-next-line no-cond-assign
  while (result = LETTERING_BOLD_REGEX.exec(content.slice(index))) {
    const before = content.slice(index, index + result.index)

    if (before) {
      chunks.push({
        type: parts.TEXT,
        content: before
      });
    }

    chunks.push({
      type: parts.LETTERING_BOLD,
      content: result[1]
    })

    index += result.index + result[0].length;
  }

  const after = content.slice(index);
  if (after) {
    chunks.push({
      type: parts.TEXT,
      content: after
    });
  }

  return chunks;
}
