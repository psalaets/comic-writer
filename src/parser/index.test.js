import parse from './index';

describe('parse()', () => {
  it('single page', () => {
    const result = parse(`Page 1`);

    expect(result).toMatchSnapshot();
  });

  it('two pages', () => {
    const result = parse(`Page 1

Page 2`);

    expect(result).toMatchSnapshot();
  });

  it('no blank lines', () => {
    const result = parse(`Page 1
Page 2`);

    expect(result).toMatchSnapshot();
  });

  it('many blank lines', () => {
    const result = parse(`Page 1





Page 2`);

    expect(result).toMatchSnapshot();
  });

  it('pages with panels', () => {
    const result = parse(`Page 1

Page 2

Panel 1

Panel 2`);

    expect(result).toMatchSnapshot();
  });

  it('pages with panels and lettering', () => {
    const result = parse(`Page 1

Page 2

Panel 1

> bob: ok

> bob: try it again

Panel 2

This is panel description.

> sfx: boom

> caption: yep`);

    expect(result).toMatchSnapshot();
  });

  it('dialogue with bold', () => {
    const result = parse(`Page 1
Panel 1

> bob: eat **this** not **that**!

> bob: this is *not*`);

    expect(result).toMatchSnapshot();
  });

  it('dialogue starts with bold', () => {
    const result = parse(`Page 1
Panel 1

> bob: **only bold**

> bob: **starts bold** and then
`);

    expect(result).toMatchSnapshot();
  });

  it('caption with bold', () => {
    const result = parse(`Page 1
Panel 1

> caption: **I** didn't mean **that**!

> caption: this is *not*`);

    expect(result).toMatchSnapshot();
  });

  it('kitchen sink', () => {
    const result = parse(`title: test
issue: 1
by: the author

Page 1

Panel 1

Some characters stand around. And do stuff.

> bob: ok

> bob: try it **again**

Panel 2

This is panel description.

> sfx: boom

> caption: yep`);

    expect(result).toMatchSnapshot();
  });
});