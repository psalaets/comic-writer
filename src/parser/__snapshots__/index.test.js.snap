// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse() caption with bold 1`] = `
Array [
  Object {
    "captionCount": 2,
    "captionWordCount": 7,
    "content": Array [
      Object {
        "captionCount": 2,
        "captionWordCount": 7,
        "content": Array [
          Object {
            "content": Array [
              Object {
                "content": "I",
                "type": "lettering-bold",
              },
              Object {
                "content": " didn't mean ",
                "type": "text",
              },
              Object {
                "content": "that",
                "type": "lettering-bold",
              },
              Object {
                "content": "!",
                "type": "text",
              },
            ],
            "id": "1.1.1",
            "modifier": null,
            "number": 1,
            "startingLine": 4,
            "type": "caption",
            "wordCount": 4,
          },
          Object {
            "content": Array [
              Object {
                "content": "this is *not*",
                "type": "text",
              },
            ],
            "id": "1.1.2",
            "modifier": null,
            "number": 2,
            "startingLine": 6,
            "type": "caption",
            "wordCount": 3,
          },
        ],
        "dialogueCount": 0,
        "dialogueWordCount": 0,
        "id": "1.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [],
        "startingLine": 2,
        "type": "panel",
      },
    ],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 1,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
]
`;

exports[`parse() dialogue starts with bold 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [
          Object {
            "content": Array [
              Object {
                "content": "only bold",
                "type": "lettering-bold",
              },
            ],
            "id": "1.1.1",
            "modifier": null,
            "number": 1,
            "speaker": "bob",
            "startingLine": 4,
            "type": "dialogue",
            "wordCount": 2,
          },
          Object {
            "content": Array [
              Object {
                "content": "starts bold",
                "type": "lettering-bold",
              },
              Object {
                "content": " and then",
                "type": "text",
              },
            ],
            "id": "1.1.2",
            "modifier": null,
            "number": 2,
            "speaker": "bob",
            "startingLine": 6,
            "type": "dialogue",
            "wordCount": 4,
          },
        ],
        "dialogueCount": 2,
        "dialogueWordCount": 6,
        "id": "1.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [
          "bob",
          "bob",
        ],
        "startingLine": 2,
        "type": "panel",
      },
    ],
    "dialogueCount": 2,
    "dialogueWordCount": 6,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 1,
    "sfxCount": 0,
    "speakers": Array [
      "bob",
      "bob",
    ],
    "startingLine": 1,
    "type": "spread",
  },
]
`;

exports[`parse() dialogue with bold 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [
          Object {
            "content": Array [
              Object {
                "content": "eat ",
                "type": "text",
              },
              Object {
                "content": "this",
                "type": "lettering-bold",
              },
              Object {
                "content": " not ",
                "type": "text",
              },
              Object {
                "content": "that",
                "type": "lettering-bold",
              },
              Object {
                "content": "!",
                "type": "text",
              },
            ],
            "id": "1.1.1",
            "modifier": null,
            "number": 1,
            "speaker": "bob",
            "startingLine": 4,
            "type": "dialogue",
            "wordCount": 4,
          },
          Object {
            "content": Array [
              Object {
                "content": "this is *not*",
                "type": "text",
              },
            ],
            "id": "1.1.2",
            "modifier": null,
            "number": 2,
            "speaker": "bob",
            "startingLine": 6,
            "type": "dialogue",
            "wordCount": 3,
          },
        ],
        "dialogueCount": 2,
        "dialogueWordCount": 7,
        "id": "1.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [
          "bob",
          "bob",
        ],
        "startingLine": 2,
        "type": "panel",
      },
    ],
    "dialogueCount": 2,
    "dialogueWordCount": 7,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 1,
    "sfxCount": 0,
    "speakers": Array [
      "bob",
      "bob",
    ],
    "startingLine": 1,
    "type": "spread",
  },
]
`;

exports[`parse() dialogue with modifier 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [
          Object {
            "content": Array [
              Object {
                "content": "go!",
                "type": "text",
              },
            ],
            "id": "1.1.1",
            "modifier": "yell",
            "number": 1,
            "speaker": "bob",
            "startingLine": 4,
            "type": "dialogue",
            "wordCount": 1,
          },
        ],
        "dialogueCount": 1,
        "dialogueWordCount": 1,
        "id": "1.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [
          "bob",
        ],
        "startingLine": 2,
        "type": "panel",
      },
    ],
    "dialogueCount": 1,
    "dialogueWordCount": 1,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 1,
    "sfxCount": 0,
    "speakers": Array [
      "bob",
    ],
    "startingLine": 1,
    "type": "spread",
  },
]
`;

exports[`parse() kitchen sink 1`] = `
Array [
  Object {
    "id": "metadata-1",
    "name": "title",
    "startingLine": 1,
    "type": "metadata",
    "value": "test",
  },
  Object {
    "id": "metadata-2",
    "name": "issue",
    "startingLine": 2,
    "type": "metadata",
    "value": "1",
  },
  Object {
    "id": "metadata-3",
    "name": "by",
    "startingLine": 3,
    "type": "metadata",
    "value": "the author",
  },
  Object {
    "captionCount": 1,
    "captionWordCount": 1,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [
          Object {
            "content": "Some characters stand around. And do stuff.",
            "id": "paragraph-1",
            "startingLine": 9,
            "type": "paragraph",
          },
          Object {
            "content": Array [
              Object {
                "content": "ok",
                "type": "text",
              },
            ],
            "id": "1.1.1",
            "modifier": null,
            "number": 1,
            "speaker": "bob",
            "startingLine": 11,
            "type": "dialogue",
            "wordCount": 1,
          },
          Object {
            "content": Array [
              Object {
                "content": "try it ",
                "type": "text",
              },
              Object {
                "content": "again",
                "type": "lettering-bold",
              },
            ],
            "id": "1.1.2",
            "modifier": null,
            "number": 2,
            "speaker": "bob",
            "startingLine": 13,
            "type": "dialogue",
            "wordCount": 3,
          },
        ],
        "dialogueCount": 2,
        "dialogueWordCount": 4,
        "id": "1.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [
          "bob",
          "bob",
        ],
        "startingLine": 7,
        "type": "panel",
      },
      Object {
        "captionCount": 1,
        "captionWordCount": 1,
        "content": Array [
          Object {
            "content": "This is panel description.",
            "id": "paragraph-2",
            "startingLine": 17,
            "type": "paragraph",
          },
          Object {
            "content": "boom",
            "id": "1.2.3",
            "modifier": null,
            "number": 3,
            "startingLine": 19,
            "type": "sfx",
          },
          Object {
            "content": Array [
              Object {
                "content": "yep",
                "type": "text",
              },
            ],
            "id": "1.2.4",
            "modifier": null,
            "number": 4,
            "startingLine": 21,
            "type": "caption",
            "wordCount": 1,
          },
        ],
        "dialogueCount": 0,
        "dialogueWordCount": 0,
        "id": "1.2",
        "number": 2,
        "sfxCount": 1,
        "speakers": Array [],
        "startingLine": 15,
        "type": "panel",
      },
    ],
    "dialogueCount": 2,
    "dialogueWordCount": 4,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 2,
    "sfxCount": 1,
    "speakers": Array [
      "bob",
      "bob",
    ],
    "startingLine": 5,
    "type": "spread",
  },
]
`;

exports[`parse() many blank lines 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "2",
    "label": "2",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 7,
    "type": "spread",
  },
]
`;

exports[`parse() no blank lines 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "2",
    "label": "2",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 2,
    "type": "spread",
  },
]
`;

exports[`parse() pages with panels 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [],
        "dialogueCount": 0,
        "dialogueWordCount": 0,
        "id": "2.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [],
        "startingLine": 5,
        "type": "panel",
      },
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [],
        "dialogueCount": 0,
        "dialogueWordCount": 0,
        "id": "2.2",
        "number": 2,
        "sfxCount": 0,
        "speakers": Array [],
        "startingLine": 7,
        "type": "panel",
      },
    ],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "2",
    "label": "2",
    "pageCount": 1,
    "panelCount": 2,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 3,
    "type": "spread",
  },
]
`;

exports[`parse() pages with panels and lettering 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
  Object {
    "captionCount": 1,
    "captionWordCount": 1,
    "content": Array [
      Object {
        "captionCount": 0,
        "captionWordCount": 0,
        "content": Array [
          Object {
            "content": Array [
              Object {
                "content": "ok",
                "type": "text",
              },
            ],
            "id": "2.1.1",
            "modifier": null,
            "number": 1,
            "speaker": "bob",
            "startingLine": 7,
            "type": "dialogue",
            "wordCount": 1,
          },
          Object {
            "content": Array [
              Object {
                "content": "try it again",
                "type": "text",
              },
            ],
            "id": "2.1.2",
            "modifier": null,
            "number": 2,
            "speaker": "bob",
            "startingLine": 9,
            "type": "dialogue",
            "wordCount": 3,
          },
        ],
        "dialogueCount": 2,
        "dialogueWordCount": 4,
        "id": "2.1",
        "number": 1,
        "sfxCount": 0,
        "speakers": Array [
          "bob",
          "bob",
        ],
        "startingLine": 5,
        "type": "panel",
      },
      Object {
        "captionCount": 1,
        "captionWordCount": 1,
        "content": Array [
          Object {
            "content": "This is panel description.",
            "id": "paragraph-1",
            "startingLine": 13,
            "type": "paragraph",
          },
          Object {
            "content": "boom",
            "id": "2.2.3",
            "modifier": null,
            "number": 3,
            "startingLine": 15,
            "type": "sfx",
          },
          Object {
            "content": Array [
              Object {
                "content": "yep",
                "type": "text",
              },
            ],
            "id": "2.2.4",
            "modifier": null,
            "number": 4,
            "startingLine": 17,
            "type": "caption",
            "wordCount": 1,
          },
        ],
        "dialogueCount": 0,
        "dialogueWordCount": 0,
        "id": "2.2",
        "number": 2,
        "sfxCount": 1,
        "speakers": Array [],
        "startingLine": 11,
        "type": "panel",
      },
    ],
    "dialogueCount": 2,
    "dialogueWordCount": 4,
    "id": "2",
    "label": "2",
    "pageCount": 1,
    "panelCount": 2,
    "sfxCount": 1,
    "speakers": Array [
      "bob",
      "bob",
    ],
    "startingLine": 3,
    "type": "spread",
  },
]
`;

exports[`parse() single page 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
]
`;

exports[`parse() two pages 1`] = `
Array [
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "1",
    "label": "1",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 1,
    "type": "spread",
  },
  Object {
    "captionCount": 0,
    "captionWordCount": 0,
    "content": Array [],
    "dialogueCount": 0,
    "dialogueWordCount": 0,
    "id": "2",
    "label": "2",
    "pageCount": 1,
    "panelCount": 0,
    "sfxCount": 0,
    "speakers": Array [],
    "startingLine": 3,
    "type": "spread",
  },
]
`;
